compiler/rustc_middle/src/ty/parameterized.rs

```diff
@@ -0,0 +1,127 @@
+}
+
+impl<T: ParameterizedOverTcx> ParameterizedOverTcx for ty::Binder<'static, T> {
+    type Value<'tcx> = ty::Binder<'tcx, T::Value<'tcx>>;
+}
+
+#[macro_export]
+macro_rules! trivially_parameterized_over_tcx {
+    ($($ty:ty),+ $(,)?) => {
+        $(
+            impl ParameterizedOverTcx for $ty {
+                #[allow(unused_lifetimes)]
+                type Value<'tcx> = $ty;
+            }
+        )*
+    }
+}
+
+trivially_parameterized_over_tcx! {
+    usize,
+    (),
+    u32,
+    std::string::String,
+    crate::metadata::ModChild,
+    crate::middle::codegen_fn_attrs::CodegenFnAttrs,
+    crate::middle::exported_symbols::SymbolExportInfo,
+    crate::mir::ConstQualifs,
+    ty::Generics,
+    ty::ImplPolarity,
+    ty::ReprOptions,
+    ty::TraitDef,
+    ty::Visibility,
+    ty::adjustment::CoerceUnsizedInfo,
+    ty::fast_reject::SimplifiedTypeGen<DefId>,
+    rustc_ast::Attribute,
+    rustc_ast::MacArgs,
+    rustc_attr::ConstStability,
+    rustc_attr::Deprecation,
+    rustc_attr::Stability,
+    rustc_hir::Constness,
+    rustc_hir::Defaultness,
+    rustc_hir::GeneratorKind,
+    rustc_hir::IsAsync,
+    rustc_hir::LangItem,
+    rustc_hir::def::DefKind,
+    rustc_hir::def_id::DefIndex,
+    rustc_hir::definitions::DefKey,
+    rustc_index::bit_set::FiniteBitSet<u32>,
+    rustc_session::cstore::ForeignModule,
+    rustc_session::cstore::LinkagePreference,
+    rustc_session::cstore::NativeLib,
+    rustc_span::DebuggerVisualizerFile,
+    rustc_span::ExpnData,
+    rustc_span::ExpnHash,
+    rustc_span::ExpnId,
+    rustc_span::SourceFile,
+    rustc_span::Span,
+    rustc_span::Symbol,
+    rustc_span::def_id::DefPathHash,
+    rustc_span::hygiene::SyntaxContextData,
+    rustc_span::symbol::Ident,
+    rustc_type_ir::Variance,
+}
+
+impl ParameterizedOverTcx for Ty<'static> {
+    type Value<'tcx> = Ty<'tcx>;
+}
+
+impl ParameterizedOverTcx for ty::FnSig<'static> {
+    type Value<'tcx> = ty::FnSig<'tcx>;
+}
+
+impl ParameterizedOverTcx for ty::GenericPredicates<'static> {
+    type Value<'tcx> = ty::GenericPredicates<'tcx>;
+}
+
+impl ParameterizedOverTcx for ty::TraitRef<'static> {
+    type Value<'tcx> = ty::TraitRef<'tcx>;
+}
+
+impl ParameterizedOverTcx for ty::Const<'static> {
+    type Value<'tcx> = ty::Const<'tcx>;
+}
+
+impl ParameterizedOverTcx for ty::Predicate<'static> {
+    type Value<'tcx> = ty::Predicate<'tcx>;
+}
+
+impl ParameterizedOverTcx for ty::GeneratorDiagnosticData<'static> {
+    type Value<'tcx> = ty::GeneratorDiagnosticData<'tcx>;
+}
+
+impl ParameterizedOverTcx for crate::mir::Body<'static> {
+    type Value<'tcx> = crate::mir::Body<'tcx>;
+}
+
+impl ParameterizedOverTcx for crate::thir::abstract_const::Node<'static> {
+    type Value<'tcx> = crate::thir::abstract_const::Node<'tcx>;
+}
+
+impl ParameterizedOverTcx for crate::middle::exported_symbols::ExportedSymbol<'static> {
+    type Value<'tcx> = crate::middle::exported_symbols::ExportedSymbol<'tcx>;
+}
```

Comment:
I imagine that this could be folded in a macro?

Reply:
So I tried doing something like

`$($ident:ident ::)* $name:ident < $lt:lifetime >` and I was getting some macro rules parse ambiguity because it's ambiguously a path.

Path segments also include their lifetimes, so I can't just do `$path:path < $lt:lifetime >` either...

Reply:
I could possibly use some other sigil to separate the path segments, but that's kinda ugly so I didn't really give it a try.

Reply:
I wouldn't expect the macro to need to take anything expect for the name of the type?

Reply:
I made a macro version that takes just a single identifier in #97376, which suffices I guess.